### PR TITLE
Order Creation: reusable component for the form's sections

### PR DESCRIPTION
**Pull request by hichamboushaba**
### Description
Since the order note's subject wasn't decided today, I decided to spend some time checking if we can improve our workflow by having a reusable view for the form's sections, which resulted in this draft PR.
I tried to keep the component simple, so it won't be able to handle the payments' section, but it would allow handling all the other sections:

Screen Shot 2021-12-03 at 16 56 02   Screen Shot 2021-12-03 at 16 56 09

The styling of the `edit` button might need some adjustments to be consistent (we need to either fix it for this component or for the status section)

If you think this kind of abstraction is not good for us, just let me know, we can drop this PR 🙂, and if you have remarks that can improve the Component's API, please share them 🙏 

If you want to play around with the component and to see how we can use it, please apply this patch:
```
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationViewModel.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationViewModel.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationViewModel.kt	(revision 96b0ed702ed2cd5411c097aae9678e5662637139)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationViewModel.kt	(date 1638547390573)
@@ -20,4 +20,8 @@
     fun onOrderStatusChanged(status: Order.Status) {
         orderDraft = orderDraft.copy(status = status)
     }
+
+    fun onNoteAdded() {
+        orderDraft = orderDraft.copy(customerNote = "New Note")
+    }
 }
Index: WooCommerce/src/main/res/layout/fragment_order_creation_form.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/res/layout/fragment_order_creation_form.xml b/WooCommerce/src/main/res/layout/fragment_order_creation_form.xml
--- a/WooCommerce/src/main/res/layout/fragment_order_creation_form.xml	(revision 96b0ed702ed2cd5411c097aae9678e5662637139)
+++ b/WooCommerce/src/main/res/layout/fragment_order_creation_form.xml	(date 1638547591046)
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <androidx.appcompat.widget.LinearLayoutCompat xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
     android:layout_width="match_parent"
     android:layout_height="match_parent"
     android:orientation="vertical">
@@ -8,6 +9,20 @@
         android:id="@+id/order_status_view"
         style="@style/Woo.Card"
         android:layout_width="match_parent"
-        android:layout_height="wrap_content"/>
+        android:layout_height="wrap_content" />
 
+    <com.woocommerce.android.ui.orders.creation.views.OrderCreationSectionView
+        android:id="@+id/customer_section"
+        style="@style/Woo.Card"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        app:header="Customer"/>
+
+    <com.woocommerce.android.ui.orders.creation.views.OrderCreationSectionView
+        android:id="@+id/note_section"
+        style="@style/Woo.Card"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        app:header="Order Note" />
+
 </androidx.appcompat.widget.LinearLayoutCompat>
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationFormFragment.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationFormFragment.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationFormFragment.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationFormFragment.kt	(revision 96b0ed702ed2cd5411c097aae9678e5662637139)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/OrderCreationFormFragment.kt	(date 1638547563894)
@@ -5,17 +5,21 @@
 import androidx.fragment.app.viewModels
 import androidx.hilt.navigation.fragment.hiltNavGraphViewModels
 import androidx.navigation.fragment.findNavController
+import com.google.android.material.textview.MaterialTextView
 import com.woocommerce.android.R
 import com.woocommerce.android.databinding.FragmentOrderCreationFormBinding
 import com.woocommerce.android.extensions.handleDialogResult
 import com.woocommerce.android.extensions.navigateSafely
 import com.woocommerce.android.extensions.takeIfNotEqualTo
+import com.woocommerce.android.model.Address
 import com.woocommerce.android.model.Order
 import com.woocommerce.android.ui.base.BaseFragment
 import com.woocommerce.android.ui.orders.OrderNavigationTarget.ViewOrderStatusSelector
 import com.woocommerce.android.ui.orders.creation.OrderCreationFormViewModel.ShowStatusTag
+import com.woocommerce.android.ui.orders.creation.views.OrderCreationSectionView
 import com.woocommerce.android.ui.orders.details.OrderDetailViewModel.OrderStatusUpdateSource
 import com.woocommerce.android.ui.orders.details.OrderStatusSelectorDialog.Companion.KEY_ORDER_STATUS_RESULT
+import com.woocommerce.android.ui.orders.details.views.OrderDetailCustomerInfoView
 import com.woocommerce.android.viewmodel.MultiLiveEvent.Event
 import dagger.hilt.android.AndroidEntryPoint
 
@@ -41,11 +45,44 @@
                 formViewModel.onEditOrderStatusSelected(sharedViewModel.currentStatus)
             }
         )
+
+        noteSection.setAddButtons(listOf(
+            OrderCreationSectionView.AddButton(
+                text = "Add Note",
+                onClickListener = {
+                    sharedViewModel.onNoteAdded()
+                }
+            )
+        ))
+        noteSection.setOnEditButtonClicked {
+            // Handle edit button
+        }
+        customerSection.setAddButtons(listOf(
+            OrderCreationSectionView.AddButton(
+                text = "Add Customer",
+                onClickListener = {
+                    // An example of we can use another custom View for managing the section's content
+                    // Should be in [setupObserversWith]
+                    val content = OrderDetailCustomerInfoView(requireContext())
+                    val tempOrder = Order.EMPTY.copy(
+                        shippingAddress = Address.EMPTY.copy(
+                            firstName = "Jhon", lastName = "Doe", country = "United States"
+                        ),
+                        billingAddress = Address.EMPTY.copy(
+                            firstName = "Jhon", lastName = "Doe", country = "United States"
+                        )
+                    )
+                    content.updateCustomerInfo(tempOrder, false, true)
+                    customerSection.updateContent(content)
+                }
+            )
+        ))
     }
 
     private fun setupObserversWith(binding: FragmentOrderCreationFormBinding) {
         sharedViewModel.orderDraftData.observe(viewLifecycleOwner) { oldOrderData, newOrderData ->
             binding.orderStatusView.updateOrder(newOrderData)
+            bindOrderNote(binding.noteSection, newOrderData.customerNote)
             newOrderData.takeIfNotEqualTo(oldOrderData?.status) {
                 formViewModel.requestStatusTagData(newOrderData.status)
             }
@@ -56,6 +93,15 @@
         }
     }
 
+    private fun bindOrderNote(section: OrderCreationSectionView, note: String) {
+        note.takeIf { it.isNotBlank() }
+            ?.let {
+                MaterialTextView(requireContext()).apply { text = note }
+            }.let { view ->
+                section.updateContent(view)
+            }
+    }
+
     private fun setupHandleResults() {
         handleDialogResult<OrderStatusUpdateSource>(
             key = KEY_ORDER_STATUS_RESULT,
```

### CODE REVIEW

**Comment by hichamboushaba** [comment]
@wzieba @ThomazFB as I'm AFK next week, please feel free to take any decision or make any code adjustments to this PR while I'm away 🙏

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/views/OrderCreationSectionView.kt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+) : MaterialCardView(ctx, attrs, defStyleAttr) {
+    private val binding = OrderCreationSectionBinding.inflate(LayoutInflater.from(ctx), this, true)
+
+    var header: CharSequence
+        get() = binding.headerLabel.text
+        set(value) {
+            binding.headerLabel.text = value
+        }
+
+    var keepAddButtons: Boolean = false
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/views/OrderCreationSectionView.kt:30]: ❓ If I understood correctly it's "keep add buttons even if content updated" right? So this is for the case of e.g. products - when adding one product shouldn't disable possibility to add another right?

**Review by wzieba — APPROVED** [review]
Nice, very neat. I love it 🙂. Thanks for preparing sample.

> If you think this kind of abstraction is not good for us, just let me know, we can drop this PR 🙂, and if you have remarks that can improve the Component's API, please share them 🙏

I think that it is a good abstraction - at this moment looks generic enough and I can't see flaws in API design. At the first glance I was wondering what's purpose of `keepAddButtons` but I think I got it (hence comment)

> The styling of the edit button might need some adjustments to be consistent (we need to either fix it for this component or for the status section)

Honestly I like `edit` button here more - ripple's color and canvas and font size is better to me. I'd opt to adjust status section - it's too wide generally IMO.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/views/OrderCreationSectionView.kt (hunk 1 above)
Comment (hichamboushaba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/views/OrderCreationSectionView.kt:30]: Yes, that's its purpose, do you think adding a comment would make it clearer?

Comment (wzieba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/creation/views/OrderCreationSectionView.kt:30]: Thanks. Looking at this today, being in context already, I think it's better to leave it without a comment.

**Comment by hichamboushaba** [comment]
Thanks for the review @wzieba, I'll merge this to make it available for the customer order note, and for products and customer sections.

**Pull request merged**